pkg/util/fedinformer/handlers_test.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2024 The Karmada Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fedinformer
+
+import (
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+type TestObject struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+	Spec string
+}
+
+func (in *TestObject) DeepCopyObject() runtime.Object {
+	return &TestObject{
+		TypeMeta:   in.TypeMeta,
+		ObjectMeta: in.ObjectMeta,
+		Spec:       in.Spec,
+	}
+}
+
+type CustomResourceEventHandler struct {
+	handler cache.ResourceEventHandler
+}
+
+func (c *CustomResourceEventHandler) OnAdd(obj interface{}, isInInitialList bool) {
+	if h, ok := c.handler.(interface{ OnAdd(interface{}, bool) }); ok {
+		h.OnAdd(obj, isInInitialList)
+	} else {
+		c.handler.OnAdd(obj, false)
+	}
+}
+
+func (c *CustomResourceEventHandler) OnUpdate(oldObj, newObj interface{}) {
+	c.handler.OnUpdate(oldObj, newObj)
+}
+
+func (c *CustomResourceEventHandler) OnDelete(obj interface{}) {
+	c.handler.OnDelete(obj)
+}
+
+func TestNewHandlerOnAllEvents(t *testing.T) {
+	var calledWith runtime.Object
+	fn := func(obj interface{}) {
+		calledWith = obj.(runtime.Object)
+	}
+
+	handler := &CustomResourceEventHandler{NewHandlerOnAllEvents(fn)}
+
+	testObj := &TestObject{ObjectMeta: metav1.ObjectMeta{Name: "test-obj"}}
+
+	handler.OnAdd(testObj, false)
+	if calledWith != testObj {
+		t.Errorf("AddFunc: expected %v, got %v", testObj, calledWith)
+	}
+
+	newObj := &TestObject{ObjectMeta: metav1.ObjectMeta{Name: "test-obj"}, Spec: "updated"}
+	handler.OnUpdate(testObj, newObj)
+	if calledWith != newObj {
+		t.Errorf("UpdateFunc: expected %v, got %v", newObj, calledWith)
+	}
+
+	handler.OnDelete(testObj)
+	if calledWith != testObj {
+		t.Errorf("DeleteFunc: expected %v, got %v", testObj, calledWith)
+	}
+
+	deletedObj := cache.DeletedFinalStateUnknown{Obj: testObj}
```

Comment:
Shouldn't this object be changed, say with newObj, to prove the test is correct?

Reply:
I think the same object testObj is valid because we're verifying that the handler doesn't modify the object during the delete process but if you say i could create a newObj for it.

Reply:
After the previous test content occurs, the target object should have changed, which does not prove that the current operation sets the correct target value.